<a href="https://colab.research.google.com/github/reguelpangrib29/prak-ai-2217051084-reguel-a/blob/main/Reguel_Andreas_Agustinus_Pangaribuan_2217051084_A_UAP_Responsi_AI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# TEMA SOAL

NPM : 2217051084 --> Akhiran angka '4' -> Tema Soal 2

Tema Soal :
2. Menentukan Pemain Bola Terbaik

  Anda adalah seorang pelatih sebuah klub sepak bola dan sedang
  menentukan pemain terbaik di klub untuk mengikuti pertandingan
  selanjutnya. Menggunakan metode AHP, bagaimana anda akan
  menentukan pemain terbaik? Gunakan kriteria dan pemain yang ingin
  anda nilai.


In [10]:
# Menerapkan Metode AHP untuk menentukan pemain bola terbaik

1. Mendefinisikan Hierarki Keputusan

*   Tujuan : Menentukan pemain terbaik klub sepak bola.
*   Kriteria : Kriteria yang relevan (Shooting Accuracy, Passing Accuracy, Defensive Skiils, Stamina).
*   Alternatif : Memilih player untuk evaluasi (Cristiano Ronaldo, Lionel Messi, Takefusa Kubo, Kim Min Jae).

2. Membuat Perbandingan Berpasangan (Pairwise Comparison) untuk kriteria :

*  Menggunakan skala 1-9 untuk mengukur preferensi berdasarkan skor prioritas utama.

3. Menghitung Bobot Kriteria dan Konsistensi.

4. Struktur AHP Hierarchy :

*   Goal : Menentukan pemain sepak bola terbaik.
*   Kriteria : Akurasi menembak, Akurasi Passing, Skill Bertahan, dan Stamina.
*   Alternatif :  Cristiano Ronaldo, Lionel Messi, Takefusa Kubo, Kim Min Jae.

5. Hitung Vektor Prioritas :

*   Memeriksa rasio konsistensi (CR) untuk memastikan bahwa penilaiannya konsisten. CR sebesar 0.1 atau kurang anggap dapat diterima.

6. Agregat Bobot :

*   Gabungkan bobot kriteria dan bobot alternatif untuk menentukan peringkat pemain secara keseluruhan.
















## Langkah 1 : Install Required Package.

In [9]:
!pip install ahpy

## Langkah 2 : Import Package.

In [17]:
import ahpy
import numpy as np
import pandas as pd

## Langkah 3 : Menentukan Matriks Perbandingan Berpasangan

*   Kriteria (C)
*   Player (P)
*   Ranking (R)



In [18]:
criteria = {
    ('Shooting Accuracy', 'Passing Accuracy'): 3,
    ('Shooting Accuracy', 'Defensive Skills'): 5,
    ('Shooting Accuracy', 'Stamina'): 7,
    ('Passing Accuracy', 'Defensive Skills'): 2,
    ('Passing Accuracy', 'Stamina'): 5,
    ('Defensive Skills', 'Stamina'): 3
}

In [19]:
# Perbandingan berpasangan untuk setiap pemain di bawah setiap kriteria
players_shooting = {
    ('Cristiano Ronaldo', 'Lionel Messi'): 2,
    ('Cristiano Ronaldo', 'Takefusa Kubo'): 5,
    ('Cristiano Ronaldo', 'Kim Min Jae'): 7,
    ('Lionel Messi', 'Takefusa Kubo'): 3,
    ('Lionel Messi', 'Kim Min Jae'): 5,
    ('Takefusa Kubo', 'Kim Min Jae'): 2
}

players_passing = {
    ('Cristiano Ronaldo', 'Lionel Messi'): 3,
    ('Cristiano Ronaldo', 'Takefusa Kubo'): 5,
    ('Cristiano Ronaldo', 'Kim Min Jae'): 6,
    ('Lionel Messi', 'Takefusa Kubo'): 4,
    ('Lionel Messi', 'Kim Min Jae'): 5,
    ('Takefusa Kubo', 'Kim Min Jae'): 2
}

players_defensive = {
    ('Cristiano Ronaldo', 'Lionel Messi'): 2,
    ('Cristiano Ronaldo', 'Takefusa Kubo'): 3,
    ('Cristiano Ronaldo', 'Kim Min Jae'): 4,
    ('Lionel Messi', 'Takefusa Kubo'): 2,
    ('Lionel Messi', 'Kim Min Jae'): 3,
    ('Takefusa Kubo', 'Kim Min Jae'): 3
}

players_stamina = {
    ('Cristiano Ronaldo', 'Lionel Messi'): 4,
    ('Cristiano Ronaldo', 'Takefusa Kubo'): 6,
    ('Cristiano Ronaldo', 'Kim Min Jae'): 8,
    ('Lionel Messi', 'Takefusa Kubo'): 2,
    ('Lionel Messi', 'Kim Min Jae'): 5,
    ('Takefusa Kubo', 'Kim Min Jae'): 3
}

## Langkah 4 : Menghitung Bobot Kriteria dan Konsistensi

In [20]:
criteria_matrix = np.array([
    [1, 3, 5, 7],
    [1/3, 1, 2, 5],
    [1/5, 1/2, 1, 3],
    [1/7, 1/5, 1/3, 1]
])

# Normalisasi Kriteria Matrix
normalized_criteria_matrix = criteria_matrix / criteria_matrix.sum(axis=0)

# Hitung vektor prioritas (bobot) untuk kriteria
criteria_weights = normalized_criteria_matrix.mean(axis=1)
print("Criteria Weights:", criteria_weights)

# Mengkalkulasikan konsistensi vektor
consistency_vector = criteria_matrix.dot(criteria_weights)
consistency_index = (consistency_vector / criteria_weights).mean() - len(criteria_weights)
random_index = {1: 0.00, 2: 0.00, 3: 0.58, 4: 0.90, 5: 1.12, 6: 1.24, 7: 1.32, 8: 1.41, 9: 1.45, 10: 1.49}
consistency_ratio = consistency_index / random_index[len(criteria_weights)]
print("Consistency Ratio (CR):", consistency_ratio)

# Periksa apakah konsistensi diterima
if consistency_ratio < 0.1:
    print("Criteria is consistent.")
else:
    print("Criteria is not consistent.")


Criteria Weights: [0.5680972  0.2410324  0.13330029 0.05757012]
Consistency Ratio (CR): 0.08676672407943897
Criteria is consistent.


## Langkah 5 : Matriks Perbandingan Berpasangan Alternatif

Untuk setiap kriteria, kita akan membuat matriks perbandingan berpasangan untuk alternatif skill player.
Berdasarkan shooting, passing, defensive dan stamina pemain :

In [21]:
# Matriks perbandingan berpasangan untuk alternatif berdasarkan shooting
players_shooting = np.array([
    [1, 2, 5, 7],
    [1/2, 1, 3, 5],
    [1/5, 1/3, 1, 2],
    [1/7, 1/5, 1/2, 1]
])

# Matriks perbandingan berpasangan untuk alternatif berdasarkan passing
players_passing = np.array([
    [1, 3, 5, 6],
    [1/3, 1, 4, 5],
    [1/5, 1/4, 1, 2],
    [1/6, 1/5, 1/2, 1]
])

# Matriks perbandingan berpasangan untuk alternatif berdasarkan defensive
players_defensive = np.array([
    [1, 2, 3, 4],
    [1/2, 1, 2, 3],
    [1/3, 1/2, 1, 3],
    [1/4, 1/3, 1/3, 1]
])

# Matriks perbandingan berpasangan untuk alternatif berdasarkan stamina
players_stamina = np.array([
    [1, 4, 6, 8],
    [1/4, 1, 2, 5],
    [1/6, 1/2, 1, 3],
    [1/8, 1/5, 1/3, 1]
])

# Menghitung kalkulasi alternatif weights dan konsistensi rasio
def calculate_weights(matrix):
    normalized_matrix = matrix / matrix.sum(axis=0)
    weights = normalized_matrix.mean(axis=1)
    consistency_vector = matrix.dot(weights)
    consistency_index = (consistency_vector / weights).mean() - len(weights)
    consistency_ratio = consistency_index / random_index[len(weights)]
    return weights, consistency_ratio

shooting_weights, shooting_cr = calculate_weights(players_shooting)
passing_weights, passing_cr = calculate_weights(players_passing)
defensive_weights, defensive_cr = calculate_weights(players_defensive)
stamina_weights, stamina_cr = calculate_weights(players_stamina)

print("Shooting Weights:", shooting_weights)
print("Shooting Consistency Ratio:", shooting_cr)
print("Passing Weights:", passing_weights)
print("Passing Consistency Ratio:", passing_cr)
print("Defensive Weights:", defensive_weights)
print("Defensive Consistency Ratio:", defensive_cr)
print("Stamina Weights:", stamina_weights)
print("Stamina Consistency Ratio:", stamina_cr)

Shooting Weights: [0.52541396 0.30086488 0.11036581 0.06335535]
Shooting Consistency Ratio: 0.022380140906502137
Passing Weights: [0.54178863 0.28972319 0.10298052 0.06550766]
Passing Consistency Ratio: 0.14792547883089824
Defensive Weights: [0.45976493 0.27234658 0.1802642  0.0876243 ]
Defensive Consistency Ratio: 0.09826603538961365
Stamina Weights: [0.6159621  0.21150103 0.11986021 0.05267665]
Stamina Consistency Ratio: 0.11454207518771081


## Langkah 6 : Menggabungkan Bobot Kriteria dan Bobot Alternatif

In [22]:
# Overall weights dari pemain
overall_weights = (
    criteria_weights[0] * shooting_weights +
    criteria_weights[1] * passing_weights +
    criteria_weights[2] * defensive_weights +
    criteria_weights[3] * stamina_weights
)

print("Overall Weights of Players:", overall_weights)

# Rank pemain dari basis overall weights
players = ['Cristiano Ronaldo', 'Lionel Messi', 'Takefusa Kubo', 'Kim Min Jae']
ranking = sorted(zip(players, overall_weights), key=lambda x: x[1], reverse=True)
print("Ranking of Players:")
for player, weight in ranking:
    print(f"{player}: {weight:.4f}")

# Menentukan pemain terbaik dengan skor tertinggi
best_alternative = ranking[0]
print(f"Best Alternative (Player with highest score): {best_alternative[0]} with score {best_alternative[1]:.4f}")

Overall Weights of Players: [0.52582262 0.28923319 0.11844979 0.06649441]
Ranking of Players:
Cristiano Ronaldo: 0.5258
Lionel Messi: 0.2892
Takefusa Kubo: 0.1184
Kim Min Jae: 0.0665
Best Alternative (Player with highest score): Cristiano Ronaldo with score 0.5258


## Kesimpulan

In [23]:
# Menggunakan dataframe
df_criteria = pd.DataFrame({
    'Criteria': ['Shooting Accuracy', 'Passing Accuracy', 'Defensive Skills', 'Stamina'],
    'Weights': criteria_weights
})

df_players = pd.DataFrame({
    'Player': players,
    'Shooting Weight': shooting_weights,
    'Passing Weight': passing_weights,
    'Defensive Weight': defensive_weights,
    'Stamina Weight': stamina_weights,
    'Overall Weight': overall_weights
})

df_ranking = pd.DataFrame(ranking, columns=['Player', 'Overall Weight'])
df_criteria
df_players
df_ranking

Player  Overall Weight
0  Cristiano Ronaldo        0.525823
1       Lionel Messi        0.289233
2      Takefusa Kubo        0.118450
3        Kim Min Jae        0.066494

Dengan menggunakan kode di atas, Kita dapat melakukan analisis AHP dengan menentukan pemain bola terbaik dari seorang pelatih klub sepak bola. Melalui rasio konsistensi (CR) untuk memastikan bahwa perbandingan berpasangan yang dilakukan adalah konsisten. Berdasarkan kriteria, maka best alternative atau alternatif terbaik dalam pemain bola terbaik adalah **Cristiano Ronaldo dengan score 0.525823 **